### PR TITLE
NO-ISSUE: extract client from reconcile context

### PR DESCRIPTION
**Pull request by danielerez**
In HypershiftAgentServiceConfigReconciler, we currently need to switch between clients (hub and spoke), since a single client property exists in AgentServiceConfigReconcileContext.

Hence, to avoid sharing the property with both clients in order to have a clear notion of the client in use; moved it to ASC struct for explicitly specifying the required client.
I.e. instead of a single ASC with alternating clients, created a separate one for the spoke using the generated client.

Also, moved the 'properties' from ASC to AgentServiceConfigReconcileContext as the map is used for sharing values between clients.

## List all the issues related to this PR

- [ ] New Feature 
- [x] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [x] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Checklist

- [x] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/controller/controllers/hypershiftagentserviceconfig_controller.go [hunk 1]
```diff
@@ -164,7 +159,7 @@ func (hr *HypershiftAgentServiceConfigReconciler) Reconcile(origCtx context.Cont
 	}
 
 	// Initialize ASC to reconcile components according to instance context
-	asc.initHASC(hr, instance)
+	asc.initHASC(hr, instance, hr.Client)
```
Comment (carbonin) [inline internal/controller/controllers/hypershiftagentserviceconfig_controller.go:162]: I know this isn't really related to this change directly, but why not have this function return a new `ASC` instance?

Initializing it here (rather than creating it) makes me feel like it is shared when the whole premise of this PR is that is it not.

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 2]
```diff
@@ -107,19 +107,23 @@ var (
...
-	client.Client
 	Log    logrus.FieldLogger
 	Scheme *runtime.Scheme
 
 	// selector and tolerations the Operator runs in and propagates to its deployments
 	NodeSelector map[string]string
 	Tolerations  []corev1.Toleration
 
+	/* properties. use this field for cross cluster communication */
+	Properties map[string]interface{}
```
Comment (carbonin) [inline internal/controller/controllers/agentserviceconfig_controller.go:118]: I'm not really sure what this is, but I don't think maps can be safely used concurrently without locking.
Maybe make this a separate PR with more clear justification?

Comment (slaviered) [reply internal/controller/controllers/agentserviceconfig_controller.go:118]: why does the map moved from HASC to the global scope?

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 3]
```diff
@@ -149,15 +155,18 @@ type ASC struct {
 	conditions *[]conditionsv1.Condition
 
 	/* properties. use this field for cross cluster communication */
-	properties map[string]interface{}
+	properties *map[string]interface{}
```
Comment (carbonin) [inline internal/controller/controllers/agentserviceconfig_controller.go:158]: The zero-value of a map is `nil` and they are generally treated as references. Why do you need a pointer here?

Here's an example if it helps: https://go.dev/play/p/l5YfudQhWT1

Comment (danielerez) [reply internal/controller/controllers/agentserviceconfig_controller.go:158]: Nice, wasn't sure about it. Fixed.

File: internal/controller/controllers/hypershiftagentserviceconfig_controller_test.go [hunk 4]
```diff
@@ -570,31 +570,31 @@ var _ = Describe("HypershiftAgentServiceConfig reconcile", func() {
 		}
 
 		It("successfully for leader election role", func() {
-			asc.initHASC(hr, hsc)
+			asc = initHASC(hr, hsc, hr.Client, ascProperties)
```
Comment (carbonin) [inline internal/controller/controllers/hypershiftagentserviceconfig_controller_test.go:573]: Shouldn't the `asc` instance be scoped to the example (i.e. remove the variable declaration and use `:=` here)? I'm imagining a test not calling `initHASC` and getting the value from the previous test run.

File: internal/controller/controllers/hypershiftagentserviceconfig_controller_test.go [hunk 5]
```diff
@@ -259,6 +259,7 @@ var _ = Describe("HypershiftAgentServiceConfig reconcile", func() {
 		mockSpokeClient = spoke_k8s_client.NewMockSpokeK8sClient(mockCtrl)
 		mockSpokeClientCache = NewMockSpokeClientCache(mockCtrl)
 
+		ascProperties = make(map[string]interface{})
```
Comment (carbonin) [inline internal/controller/controllers/hypershiftagentserviceconfig_controller_test.go:262]: If this is only relevant in an inner context then it should be defined and set in that context.

Moving it further away from where it's used makes the tests harder to understand.

File: internal/controller/controllers/hypershiftagentserviceconfig_controller.go [hunk 6]
```diff
@@ -163,8 +161,11 @@ func (hr *HypershiftAgentServiceConfigReconciler) Reconcile(origCtx context.Cont
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
+	// Create a map for properties that are shared between ASC objects
+	ascProperties := make(map[string]interface{})
```
Comment (carbonin) [inline internal/controller/controllers/hypershiftagentserviceconfig_controller.go:165]: Is moving this out solving a bug of some sort? This really feels like it deserves its own PR and an issue, or at least a separate commit.

**Review by carbonin — APPROVED** [review]
